### PR TITLE
ENHANCEMENT Enable UploadField.js within non-react sections	

### LINKED ISSUE #297: API Enable react UploadField.js support within non-react sections
In order to fully deprecate framework UploadField it will be necessary to build an uploadfield which works outside of react.

I propose creating a react wrapper around https://github.com/silverstripe/silverstripe-asset-admin/issues/296 which will bootstrap react / redux around UploadField.js, allowing it to be used in place of any uploadfield.

Acceptance Criterea:
 - The asset-admin UploadField can be used in php forms
 - This UploadField allows files to be added or removed from a has_one or many_many
 - Look and feel matches designs
 
Tasks:
 - [x] Fix js errors https://github.com/silverstripe/silverstripe-asset-admin/pull/329
 - [x] Merge framework PR https://github.com/silverstripe/silverstripe-framework/pull/6392
 - [ ] Merge asset-admin PR https://github.com/silverstripe/silverstripe-asset-admin/pull/331
 - [x] Get UploadField.js to render in a normal form
 - [x] Form submission works
 - [x] Upgrading notes / guide

### PR DESCRIPTION
**Pull request by tractorcow**
Relies on fixes in https://github.com/silverstripe/silverstripe-asset-admin/pull/329, so rebased on top of this commit.

Requires https://github.com/silverstripe/silverstripe-framework/pull/6392 to be merged first.

Fixes #297

This is what the field looks like in cms main:

image

### CODE REVIEW

File: client/src/components/UploadField/UploadField.js [hunk 1]
```diff
@@ -140,6 +140,7 @@ class UploadField extends SilverStripeComponent {
    */
   handleSelect(event) {
     event.preventDefault();
+    return false;
```
Comment (dhensby) [inline client/src/components/UploadField/UploadField.js]: My understanding is that `return false` is bad practice.

If you want to stop propagation of the event, then use `event.stopPropagation()` and not `return false`.

File: client/src/entwine/UploadField/UploadFieldShiv.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+   * Get props to use for this object
+   *
+   * @returns {*}
+   */
+  getUploadProps() {
+    return Object.assign(
+      {},
+      this.props.schema,
+      this.props.state,
+      { data: Object.assign({}, this.props.schema.data || {}, this.props.state.data || {}) }
```
Comment (flamerohr) [inline client/src/entwine/UploadField/UploadFieldShiv.js]: This should use `merge.recursive(true, {}, data, data);` so you don't have different functionality to `FormBuilder.mergeFieldData()`

Actually, I prefer if this prop could be done in the entwine `getAttributes` and take advantage of `$.extends(true, {}, data, data);` and you won't need the shiv class layer.

**Comment by tractorcow** [comment]
Feedback implemented

**Comment by dhensby** [comment]
leaving to @flamerohr to make sure he's happy with resolutions

**Comment by flamerohr** [comment]
Looks good, I had amended a: `export default fieldHolder(connect(mapStateToProps, mapDispatchToProps)(UploadField));` to 
`export default fieldHolder(ConnectedUploadField);` to reduce code duplication which I hadn't noticed in my previous review :)

File: client/src/entwine/UploadField/UploadFieldEntwine.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Shiv for inserting react UploadField into entwine forms
+ */
+jQuery.entwine('ss', ($) => {
+
+  /**
+   * See boot/index.js for `.react-boot` bootstrap
+   */
+  $('.react-boot input.entwine-uploadfield').entwine({
+
+    onunmatch() {
```
Comment (chillu) [inline client/src/entwine/UploadField/UploadFieldEntwine.js]: Did you omit `this._super();` in `onmatch()` and `onunmatch()` on purpose here?

File: client/src/entwine/UploadField/UploadFieldEntwine.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    onmatch() {
+      this.refresh();
+    },
+
+    refresh() {
+      const store = window.ss.store;
+      const props = this.getAttributes();
+      ReactDOM.render(
+        <Provider store={store}>
```
Comment (chillu) [inline client/src/entwine/UploadField/UploadFieldEntwine.js]: At first I thought that multiple `<Provider>` calls in with the same store would be bad, but it [looks like](https://github.com/reactjs/react-redux/blob/master/src/components/Provider.js) it's OK - just a shortcut for react context, the meat of the implementation is in the individual `connect()` calls

File: client/src/entwine/UploadField/UploadFieldEntwine.js (hunk 3 above)
Comment (tractorcow) [inline client/src/entwine/UploadField/UploadFieldEntwine.js]: No, I legit forgot them. I'll restore them.

**Pull request merged**